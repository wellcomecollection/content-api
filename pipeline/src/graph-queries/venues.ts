const query = `{
  collection-venue {
    title
    monday
    tuesday
    wednesday
    thursday
    friday
    saturday
    sunday
    modifiedDayOpeningTimes
  }
}`;

export default query;
