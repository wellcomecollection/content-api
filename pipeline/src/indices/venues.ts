export const index = "venues";

export const mappings = {
  dynamic: "strict",
  properties: {
    id: {
      type: "keyword",
    },
    display: {
      type: "object",
      enabled: false,
    },
    data: {
      type: "object",
      properties: {
        exceptionalClosedDays: {
          properties: {
            endDateTime: {
              type: "text",
            },
            overrideDate: {
              type: "date",
            },
            startDateTime: {
              type: "text",
            },
            type: {
              type: "text",
            },
          },
        },
        regularOpeningDays: {
          properties: {
            closes: {
              type: "text",
            },
            dayOfWeek: {
              type: "text",
            },
            isClosed: {
              type: "boolean",
            },
            opens: {
              type: "text",
            },
          },
        },
      },
    },
    filter: {
      properties: {
        id: {
          type: "keyword",
        },
        title: {
          type: "keyword",
        },
      },
    },
  },
} as const;
