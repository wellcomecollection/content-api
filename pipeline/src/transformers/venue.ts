import { asDate, TimestampField } from '@prismicio/client';

import {
  DayOfWeek,
  ElasticsearchVenue,
  ExceptionalClosedDay,
  RegularOpeningDay,
} from '@weco/content-common/types/venue';
import {
  PrismicExceptionalOpeningDays,
  PrismicRegularOpeningDay,
  VenuePrismicDocument,
} from '@weco/content-pipeline/src/types/prismic/venues';

export const transformVenue = (
  document: VenuePrismicDocument
): ElasticsearchVenue => {
  const {
    data: {
      title,
      monday,
      tuesday,
      wednesday,
      thursday,
      friday,
      saturday,
      sunday,
      modifiedDayOpeningTimes,
    },
    id,
  } = document;

  const formatRegularOpeningDay = (
    day: DayOfWeek,
    openingTimes: PrismicRegularOpeningDay
  ): RegularOpeningDay => {
    const formatTime = (time: TimestampField | undefined): string => {
      return time
        ? `${asDate(time).getHours()}:${String(
            asDate(time).getMinutes()
          ).padStart(2, '0')}`
        : '00:00';
    };

    return {
      dayOfWeek: day,
      opens: formatTime(openingTimes[0]?.startDateTime),
      closes: formatTime(openingTimes[0]?.endDateTime),
      isClosed: !openingTimes[0]?.startDateTime,
    };
  };

  const formatExceptionalClosedDays = (
    modifiedDayOpeningTimes: PrismicExceptionalOpeningDays
  ): ExceptionalClosedDay[] => {
    return modifiedDayOpeningTimes.map(day => {
      if (!asDate(day.overrideDate)) {
        throw new Error('Date for modified opening time is not valid');
      }

      return {
        overrideDate: asDate(day.overrideDate)?.toISOString(),
        type: day.type,
        startDateTime: '00:00',
        endDateTime: '00:00',
      };
    });
  };

  const regularOpeningDays = [
    formatRegularOpeningDay('monday', monday),
    formatRegularOpeningDay('tuesday', tuesday),
    formatRegularOpeningDay('wednesday', wednesday),
    formatRegularOpeningDay('thursday', thursday),
    formatRegularOpeningDay('friday', friday),
    formatRegularOpeningDay('saturday', saturday),
    formatRegularOpeningDay('sunday', sunday),
  ];

  const exceptionalClosedDays = formatExceptionalClosedDays(
    modifiedDayOpeningTimes
  );

  return {
    id,
    display: {
      type: 'Venue',
      id,
      title,
      regularOpeningDays,
      exceptionalClosedDays,
    },
    data: {
      regularOpeningDays,
      exceptionalClosedDays,
    },
    filter: {
      title: [
        title,
        title
          .toLowerCase()
          .replace(/[èéêë]/g, 'e')
          .replace(/\s+/g, '-'),
      ],
      id,
    },
  };
};
