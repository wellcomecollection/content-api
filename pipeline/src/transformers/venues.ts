import { PrismicDocument, TimestampField, asDate } from "@prismicio/client";
import {
  ElasticsearchEventDocument,
  ElasticsearchVenue,
} from "../types/transformed";
import {
  isFilledLinkToDocumentWithData,
  isImageLink,
  asText,
  asTitle,
  isNotUndefined,
} from "../helpers/type-guards";
import {
  linkedDocumentIdentifiers,
  transformSeries,
  getNextOpeningDates,
  addDays,
} from "./utils";
import {
  VenuePrismicDocument,
  RegularOpeningDay,
  ExceptionalOpeningDays,
} from "../types/prismic/venues";
import {
  DayOfWeek,
  DisplayRegularOpeningDay,
  DisplayExceptionalOpeningDays,
} from "../types/transformed/venue";
import { getdoc } from "../test-query";

import * as prismic from "@prismicio/client";

const libraryDoc = {
  id: "WsuS_R8AACS1Nwlx",
  uid: null,
  url: null,
  type: "collection-venue",
  href: "https://wellcomecollection.cdn.prismic.io/api/v2/documents/search?ref=Ze7JVhIAACQAvUOb&q=%5B%5B%3Ad+%3D+at%28document.id%2C+%22WsuS_R8AACS1Nwlx%22%29+%5D%5D",
  tags: ["ShortNoticeClosure"],
  first_publication_date: "2018-04-09T16:21:23+0000",
  last_publication_date: "2024-03-08T13:08:54+0000",
  slugs: ["library", "read-about-the-library"],
  linked_documents: [],
  lang: "en-gb",
  alternate_languages: [],
  data: {
    title: "Library",
    monday: [
      {
        startDateTime: "2022-01-23T10:00:00+0000",
        endDateTime: "2022-01-23T18:00:00+0000",
      },
    ],
    tuesday: [
      {
        startDateTime: "2022-01-23T10:00:00+0000",
        endDateTime: "2022-01-23T18:00:00+0000",
      },
    ],
    wednesday: [
      {
        startDateTime: "2022-01-23T10:00:00+0000",
        endDateTime: "2022-01-23T18:00:00+0000",
      },
    ],
    thursday: [
      {
        startDateTime: "2022-01-23T10:00:00+0000",
        endDateTime: "2022-01-23T20:00:00+0000",
      },
    ],
    friday: [
      {
        startDateTime: "2022-01-23T10:00:00+0000",
        endDateTime: "2022-01-23T18:00:00+0000",
      },
    ],
    saturday: [
      {
        startDateTime: "2022-01-23T10:00:00+0000",
        endDateTime: "2022-01-23T16:00:00+0000",
      },
    ],
    sunday: [{ startDateTime: null, endDateTime: null }],
    modifiedDayOpeningTimes: [
      {
        overrideDate: "2023-12-21T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-22T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-23T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-24T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-25T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-26T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-27T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-28T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-29T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-30T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2023-12-31T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2024-01-01T00:00:00+0000",
        type: "Christmas and New Year",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2024-03-29T00:00:00+0000",
        type: "Easter",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2024-03-30T00:00:00+0000",
        type: "Easter",
        startDateTime: null,
        endDateTime: null,
      },
      {
        overrideDate: "2024-03-31T23:00:00+0000",
        type: "Easter",
        startDateTime: null,
        endDateTime: null,
      },
    ],
  },
} as VenuePrismicDocument;

export const transformVenue = async (
  document: VenuePrismicDocument
): Promise<ElasticsearchVenue> => {
  const {
    data: {
      title,
      monday,
      tuesday,
      wednesday,
      thursday,
      friday,
      saturday,
      sunday,
      modifiedDayOpeningTimes,
    },
    id,
  } = document;

  const formatRegularOpeningDay = (
    day: DayOfWeek,
    openingTimes: RegularOpeningDay
  ): DisplayRegularOpeningDay => {
    const formatTime = (time: TimestampField | undefined): string => {
      return time
        ? `${asDate(time).getHours()}:${String(
            asDate(time).getMinutes()
          ).padStart(2, "0")}`
        : "00:00";
    };

    return {
      dayOfWeek: day,
      opens: formatTime(openingTimes[0]?.startDateTime),
      closes: formatTime(openingTimes[0]?.endDateTime),
      isClosed: !openingTimes[0]?.startDateTime,
    };
  };

  const regularOpeningDays = [
    formatRegularOpeningDay("monday", monday),
    formatRegularOpeningDay("tuesday", tuesday),
    formatRegularOpeningDay("wednesday", wednesday),
    formatRegularOpeningDay("thursday", thursday),
    formatRegularOpeningDay("friday", friday),
    formatRegularOpeningDay("saturday", saturday),
    formatRegularOpeningDay("sunday", sunday),
  ];

  // const nextOpeningDates: any = getNextOpeningDates(regularOpeningDays, modifiedDayOpeningTimes)

  console.log(getNextOpeningDates(regularOpeningDays, modifiedDayOpeningTimes));

  return {
    id,
    display: {
      type: "Venue",
      id,
      title: title,
      regularOpeningDays,
      exceptionalOpeningDays:
        modifiedDayOpeningTimes as DisplayExceptionalOpeningDays[],
    },
    query: {
      title: title,
      id,
    },
    nextOpeningDates: [{ open: new Date(), close: new Date() }],
  };
};

transformVenue(libraryDoc);
