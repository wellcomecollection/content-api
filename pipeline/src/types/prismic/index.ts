import * as prismic from "@prismicio/client";
import { PrismicImage } from "..";

/**
 * This is a convenience type for what the generic DataInterface type extend in @prismicio/client
 */
export type DataInterface = Record<
  string,
  prismic.AnyRegularField | prismic.GroupField | prismic.SliceZone
>;

/**
 * This allows us to get the DataInterface from PrismicDocuments when we
 * Need them for `ContentRelationshipField`s e.g.
 *
 *     type Doc = PrismicDocument<{ title: prismic.RichTextField }>
 *     type DataInterface = InferDataInterface<Doc> // { title: RichTextField }
 *     ContentRelationshipField<'formats', 'en-gb', DataInterface>
 *
 */
export type InferDataInterface<T> = T extends prismic.PrismicDocument<
  infer DataInterface
>
  ? DataInterface
  : never;

type Promo = { caption: prismic.RichTextField; image: PrismicImage };
export type PromoSliceZone = prismic.SliceZone<
  prismic.Slice<"editorialImage", Promo>
>;

export type CommonPrismicFields = {
  title: prismic.RichTextField;
  promo: PromoSliceZone;
};
