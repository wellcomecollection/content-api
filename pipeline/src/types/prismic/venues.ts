import * as prismic from "@prismicio/client";

export type RegularOpeningDay = prismic.GroupField<{
  startDateTime: prismic.TimestampField | null;
  endDateTime: prismic.TimestampField | null;
}>;

export type ExceptionalOpeningDays = prismic.GroupField<{
  overrideDate: prismic.TimestampField;
  type: string | null;
  startDateTime: prismic.TimestampField | null;
  endDateTime: prismic.TimestampField | null;
}>;

export type VenuePrismicDocument = prismic.PrismicDocument<{
  title: string;
  monday: RegularOpeningDay;
  tuesday: RegularOpeningDay;
  wednesday: RegularOpeningDay;
  thursday: RegularOpeningDay;
  friday: RegularOpeningDay;
  saturday: RegularOpeningDay;
  sunday: RegularOpeningDay;
  modifiedDayOpeningTimes: ExceptionalOpeningDays;
}>;
