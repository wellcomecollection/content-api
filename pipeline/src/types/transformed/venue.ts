export type Venue = {
  type: "Venue";
  id: string;
  title: string;
  regularOpeningDays: {
    dayOfWeek: DayOfWeek;
    opens: string;
    closes: string;
    isClosed?: boolean;
  }[];
  exceptionalOpeningDays: {
    overrideDate: Date;
    opens: string;
    closes: string;
    isClosed?: boolean;
  }[];
};

type DayOfWeek =
  | "monday"
  | "tuesday"
  | "wesdnesday"
  | "thursday"
  | "friday"
  | "saturday"
  | "sunday";
