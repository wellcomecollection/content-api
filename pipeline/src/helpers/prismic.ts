import * as prismic from "@prismicio/client";
import { ContentType } from "../types";

const graphQueryArticles = `{
    articles {
      title
      format {
        title
      }
      promo
      contributors {
        ...contributorsFields
        role {
          title
        }
        contributor {
          ... on people {
            name
          }
          ... on organisations {
            name
          }
        }
      }
    }
  }`.replace(/\n(\s+)/g, "\n"); // Pre-emptive removal of whitespaces as requests to the Prismic Rest API are limited to 2048 characters.

// TODO Review what prismic function we use here.
// getAllByType (no pagination blocker but only one type at a time)
// or getByType (which allows more than one type at a time).
// Review if predicates are still required
type GetPrismicDocumentsParams = {
  client: prismic.Client;
  contentTypes: ContentType[];
  after?: string;
};

export type PrismicPage<T> = {
  docs: T[];
  lastDocId?: string;
};

export const getPrismicDocuments = async <T>({
  client,
  contentTypes,
  after,
}: GetPrismicDocumentsParams): Promise<PrismicPage<T>> => {
  const docs = await client.getByType(contentTypes[0], {
    graphQuery: graphQueryArticles,
    predicates: [prismic.predicate.any("document.type", contentTypes)],
    orderings: {
      field: "document.last_publication_date",
      direction: "desc",
    },
    pageSize: 100,
    after,
  });

  const results = docs.results;
  const lastDoc = results[results.length - 1];
  const lastDocId = lastDoc?.id;

  return {
    docs: results as T[],
    lastDocId,
  };
};
