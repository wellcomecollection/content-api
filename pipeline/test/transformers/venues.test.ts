import { forEachPrismicSnapshot } from "../fixtures/prismic-snapshots";
import { transformVenue } from "../../src/transformers/venue";
import { VenuePrismicDocument } from "../../src/types/prismic/venues";

describe("eventDocument transformer", () => {
  forEachPrismicSnapshot<VenuePrismicDocument>("collection-venue")(
    "transforms venues from Prismic to the expected format",
    (prismicDocument) => {
      jest.useFakeTimers().setSystemTime(new Date("2024-03-14T00:00:00.000Z"));
      const transformed = transformVenue(prismicDocument);
      expect(transformed).toMatchSnapshot();
    }
  );
});
