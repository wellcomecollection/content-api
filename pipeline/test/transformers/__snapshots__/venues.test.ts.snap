// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventDocument transformer transforms venues from Prismic to the expected format (document: collection-venue/WsuS_R8AACS1Nwlx) 1`] = `
{
  "data": {
    "exceptionalClosedDays": [
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-03T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-05T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-24T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-26T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-08-23T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-08-25T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-12-23T00:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Christmas and New Year",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-12-24T00:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Christmas and New Year",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-12-30T00:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Christmas and New Year",
      },
    ],
    "regularOpeningDays": [
      {
        "closes": "18:00",
        "dayOfWeek": "monday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "18:00",
        "dayOfWeek": "tuesday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "18:00",
        "dayOfWeek": "wednesday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "20:00",
        "dayOfWeek": "thursday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "18:00",
        "dayOfWeek": "friday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "16:00",
        "dayOfWeek": "saturday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "00:00",
        "dayOfWeek": "sunday",
        "isClosed": true,
        "opens": "00:00",
      },
    ],
  },
  "display": {
    "exceptionalClosedDays": [
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-03T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-05T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-24T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-05-26T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-08-23T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-08-25T23:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Bank holiday",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-12-23T00:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Christmas and New Year",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-12-24T00:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Christmas and New Year",
      },
      {
        "endDateTime": "00:00",
        "overrideDate": "2024-12-30T00:00:00.000Z",
        "startDateTime": "00:00",
        "type": "Christmas and New Year",
      },
    ],
    "id": "WsuS_R8AACS1Nwlx",
    "regularOpeningDays": [
      {
        "closes": "18:00",
        "dayOfWeek": "monday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "18:00",
        "dayOfWeek": "tuesday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "18:00",
        "dayOfWeek": "wednesday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "20:00",
        "dayOfWeek": "thursday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "18:00",
        "dayOfWeek": "friday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "16:00",
        "dayOfWeek": "saturday",
        "isClosed": false,
        "opens": "10:00",
      },
      {
        "closes": "00:00",
        "dayOfWeek": "sunday",
        "isClosed": true,
        "opens": "00:00",
      },
    ],
    "title": "Library",
    "type": "Venue",
  },
  "filter": {
    "id": "WsuS_R8AACS1Nwlx",
    "title": [
      "Library",
      "library",
    ],
  },
  "id": "WsuS_R8AACS1Nwlx",
}
`;
